### ISSUE TITLE
Problem with contrastive loss in pretrain stage

### ISSUE THREAD

**Issue by tien-ngnvan** [texttron/tevatron#94]
Thanks for your great work. I meet the problem when using the same hyperparameters in [NQ example](https://github.com/texttron/tevatron/tree/main/examples/coCondenser-nq#training-with-hard-negatives) pre-train on the second stage like coCondenser (we call uptrain stage with contrastive loss). Our template includes 1 query, 1 positive and 10 negative passages with our custom dataloader using a streaming mode dataset (dataset includes two languages with 25M triplet samples), our model based on [bert-base-multilingual-cased](https://huggingface.co/bert-base-multilingual-cased) has been continuing pretrain with MLM [loss curve](https://wandb.ai/tien-ngnvan/pretrain-mBERT/reports/train-encoder_loss-23-11-07-14-12-20---Vmlldzo1ODk0MDM3). It seems pre-train on [contrastive loss](https://github.com/texttron/tevatron/blob/0e939457444f78284ab0471da74a0c74bc76a833/src/tevatron/loss.py#L7) can not be converged, here is the training script

    python -m torch.distributed.launch --nproc_per_node=8 -m asymmetric.train \
        --model_name_or_path 'asymmetric/checkpoint-10000' \
        --streaming \
        --output $saved_path \
        --do_train \
        --train_dir 'data/train' \
        --max_steps 10000 \
        --per_device_train_batch_size 32 \
        --dataset_num_proc 2 \
        --train_n_passages 8 \
        --gc_q_chunk_size 8 \
        --gc_p_chunk_size 8 \
        --untie_encoder \
        --negatives_x_device \
        --learning_rate 5e-4 \
        --weight_decay 1e-2 \
        --warmup_ratio 0.1 \
        --save_steps 1000 \
        --save_total_limit 20 \
        --logging_steps 50 \
        --q_max_len 128 \
        --p_max_len 384 \
        --fp16 \
        --report_to 'wandb' \
        --overwrite_output_dir
![image](https://github.com/texttron/tevatron/assets/98959709/b35a2b4c-9990-442a-9a98-6f64c0b99703)

**Comment by luyug** [comment]
To understand this better, can you elaborate on what hardware is used?
